packages/gatsby/src/commands/serve.js

```diff
@@ -40,6 +40,24 @@ const readMatchPaths = async program => {
   return JSON.parse(rawJSON)
 }
 
+/**
+ * Express middleware that adds url trailing slash in order to
+ * trick static middleware and thus preventing redirect
+ * This needs to be used before express static middleware
+ * @param {*} req
+ * @param {*} res
+ * @param {*} next
+ */
+const directoryRedirectPrevention = (req, res, next) => {
+  if (req.url == `/` || req.url.indexOf(`.`) || req.url.slice(-1) !== `/`)
```

Comment:
Just something that caught my eye;

- `req.url.indexOf(`.`)` will only be falsy if the dot was the first character, otherwise it's always truthy since `indexOf` will return `-1` if the needle was not found.

I suspect you want to use `.includes('.')` here, instead. Considering this is very likely not what you wanted, I wonder how thoroughly this was tested and perhaps whether tests could be added for this. (Semantically, I wonder what the point is of checking for `'.'` in the first place. Can you please explain the idea behind that heuristic?)

- `req.url.slice(-1) !== '/'` Why are you returning early if the last character is not a forward slash?

Isn't that exactly what you want to fix? This seems a bit inverted.

- We use strict comparisons (`===` instead of `==`)

- We use brackets for multi-line statements (so `if (...) { ... }` unless everything is on the same line)

So basically, I wonder what cases are actually fixed by this code. Could you post an example of what this fixed for you?

Will leave the actual review for others mentioned in the PR.

Reply:
@pvdz thank you for catching that, indeed I meant to use `.includes()` instead of `.indexOf()`.
At the same time, you missed the entire issue and the solution I presented, Please refer to Issue #19543 and PR #19567. I really don't know how to explain this further.
I will go on and create a more descriptive code, one that follows the conduct you mentioned, containing the `includes` fix as well.